library/core/src/mem/maybe_uninit.rs

```diff
@@ -573,6 +574,27 @@ impl<T> MaybeUninit<T> {
         }
     }
 
+    /// Drops the contained value in place.
+    ///
+    /// If you have ownership of the `MaybeUninit`, you can use [`assume_init`] instead.
+    ///
+    /// # Safety
+    ///
+    /// Calling this when the content is not yet fully initialized causes undefined
+    /// behavior: it is up to the caller to guarantee that the `MaybeUninit<T>` really
+    /// is in an initialized state.
```

Comment:
This unfortunately runs into a weakness of our terminology: in the `MaybeUninit` docs, we use "initialized" to refer to [the validity invariant](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html)... but in this paragraph here, it refers to the *safety invariant*.

I am not sure how to best resolve this until the terminology question is settled (https://github.com/rust-lang/unsafe-code-guidelines/issues/95). Maybe say something like "is in an initialized state and all its library-depended invariants are satisfied" or so?

Reply:
@m-ou-se any thoughts on this?

Reply:
Not yet. I'll read that blogpost first. Some personal stuff came up. Will get back to this.

Reply:
Okay, no worries. :) Just ping me when you need another round of review.

Reply:
The `# Safety` section here is the same as for [`assume_init_ref`](https://doc.rust-lang.org/nightly/core/mem/union.MaybeUninit.html#method.assume_init_ref) and [`read`](https://doc.rust-lang.org/nightly/core/mem/union.MaybeUninit.html#method.read) right now. I guess you're saying that those are also wrong/incomplete?

The [`assume_init`](https://doc.rust-lang.org/nightly/core/mem/union.MaybeUninit.html#method.assume_init) documentation does have the part about safety invariants. Maybe the others should refer to that explanation instead of copying it? Or maybe an explanation in the type level docs instead?

Reply:
For `read` itself to be well-defined (not cause UB), it is sufficient to consider validity. The same is true for `assume_init`. For the latter we added a remark, which we should likely reference (but probably not copy to avoid duplication) from the other `assume_init` methods: it is dangerous to use this to create something that is "valid" but does not satisfy all library invariant, but if you are really careful in what you do with this data (i.e., you never call any library method, you jsut move the data around), you are fine.

This is very different from `drop`, which *calls library-controlled code on the data*, namely `T`'s drop implementation. That's why the extra library invariants are much more important for `drop` than for any other method.

Does this make sense? It's hard to explain these things without having solid established terminology, so if I should add some examples or so, let me know.

Reply:
Ah! Yes, that makes sense. Thanks for explaining.

Reply:
Updated!